File: stock/stock.c
#include <stdio.h>

// gcc stock.c -O1 -fsanitize=address -fno-omit-frame-pointer -std=gnu99; ./a.out


int maxProfit(int* prices, int pricesSize){
     /* Given an array prices where prices[i] is the
      * price of a given stock on the ith day, returns
      * the maximum profit you can achieve from a
      * single transaction.
     */
    int current = 0, nxt = 1;
    int profit = 0, maximumProfit = 0;
    while(nxt < pricesSize){
        if(prices[current] < prices[nxt]){
            profit = prices[nxt] - prices[current];
            if(profit > maximumProfit){
                maximumProfit = profit;
            }
        }
        else {current = nxt;}
        nxt++;
    }
    return maximumProfit;
}


// TESTS
int main(void){
    int returnVal = 0;

    // assert [7, 1, 5, 3, 6, 4] == 5
    int p[6] = {7, 1, 5, 3, 6, 4};
    int s = sizeof(p) / sizeof(int);
    int result = maxProfit(p, s);
    printf("\n%d", result);
    if(result != 5){returnVal = 1;}

    printf("\n");
    return returnVal;
}

File: stock/stock.py
#!/usr/bin/env python3
"""stock.py"""
from typing import List
from timeit import timeit


class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        """Given an array prices where prices[i] is the
        price of a given stock on the ith day, returns
        the maximum profit you can achieve from a
        single transaction."""
        low = prices[0]
        high = low
        buy, sell, day = 1, 1, 1
        for day, price in enumerate(prices[1:], start=1):
            if price < low:
                low, high = price, price
                buy = day
            if price > high:
                high = price
                sell = day
        if buy < sell:
            return high - low
        return 0

    def bettermaxProfit(self, prices: List[int]) -> int:
        """Given an array prices where prices[i] is the
        price of a given stock on the ith day, returns
        the maximum profit you can achieve from a
        single transaction."""
        current, nxt, max_profit = 0, 1, 0
        while nxt < len(prices):
            if prices[current] < prices[nxt]:
                profit = prices[nxt] - prices[current]
                if profit > max_profit:
                    max_profit = profit
            else:
                current = nxt
            nxt += 1
        return max_profit


assert Solution().maxProfit([7, 1, 5, 3, 6, 4]) == 5
assert Solution().maxProfit([7, 6, 4, 3, 1]) == 0
assert Solution().bettermaxProfit([7, 1, 5, 3, 6, 4]) == 5
assert Solution().bettermaxProfit([7, 6, 4, 3, 1]) == 0

X = timeit(lambda: Solution().maxProfit([7, 1, 5, 3, 6, 4]), number=1000000)
print(X)

Y = timeit(lambda: Solution().bettermaxProfit([7, 1, 5, 3, 6, 4]),
           number=1000000)
print(Y)
